### ISSUE TITLE
Should `run()` exist?

### ISSUE THREAD

**Issue by Mr0grog** [segmentio/nightmare#575]
This is spawned from discussion on #501. @rosshinkley notes:

> `.run()` is really intended for internal use and usage isn't directly supported. Should it be?
1. `run` is not documented, it’s not the “official” API, and nothing internal ever calls it except `end`, which could just as easily call `then` instead. If `run` isn’t intended for external usage, it should be removed (or at least deprecated by logging warnings to the console when called until v3 is released).
2. On the other hand, even though it’s not documented, “official,” or public, I get the impression that it _is_ widely used. It comes up in issues a lot. Does that mean it should _become_ documented and official?

I can see a lot of value in simply canonizing a thing lots of people seem to be using. That said I personally like a smaller API surface, so I think I lean towards deprecating `run`. In the end, though, I’m not sure I have a real horse in the race here. I’d love to hear others’ thoughts.

**Comment by rosshinkley** [comment]
> run is not documented, it’s not the “official” API, and nothing internal ever calls it except end.

At the possible expense of being "that guy," `.then()` uses [`.run()` internally too](https://github.com/segmentio/nightmare/blob/master/lib/nightmare.js#L357).

> I can see a lot of value in simply canonizing a thing lots of people seem to be using.

Adding some historical context:  I believe the PhantomJS implementation [used `.run()` to execute the Nightmare chain](https://github.com/segmentio/nightmare/tree/1.8.2#runcb).  I would hazard a guess that a fair number of things out there were either developed against 1.x or were developed looking at examples from 1.x.

> ...  so I think I lean towards deprecating run.

Making `.run()` a method that wraps an internal `run()` method with some deprecation warnings would be something I could get behind.  Thoughts?

**Comment by Mr0grog** [comment]
> At the possible expense of being "that guy," .then() uses .run() internally too.

Oh, fine! ;)

> I believe the PhantomJS implementation used .run() to execute the Nightmare chain.

Ah! Well that explains a lot. That actually makes me feel not-so-hot about deprecating it, too :\

**Comment by ripper234** [comment]
@rosshinkley writes in #979 

> TL;DR: don't use .run(). It's a holdover from Nightmare 1.x. Use .then() instead.

If that's indeed the case, let's decide this & close this issue.

My 2 cents, without getting into the whole history here, is the syntax of `run()` is nice, and I don't really understand why it can't/shouldn't be a wrapper. But whether it is supported or isn't supported, a decision one way or another would be nice :)

**Comment by casesandberg** [comment]
We are going to stick with `.then()` as it is the canonical syntax for async operations with promises in JS.

**Comment by Mr0grog** [comment]
…shouldn’t this result in a PR to implement that, then? Not just closing it?

**Comment by casesandberg** [comment]
My bad, I had an overzealous finger, I am all for deprecating it! We are working on the headless chrome implementation for v3 and it removes it there https://github.com/segmentio/nightmare/blob/v3/lib/actions.js

**Comment by Mr0grog** [comment]
Ah, ok.
